Add first 3 client tests, for find_content and choosing private key option

KolbyML opened the pull request:
~~So I thought of what I thought was a good way to add 3 client tests. I didn't want to add them in main.rs since I thought 2 and 3 client tests should be seperated as to keep them cleanly seperated. Also to improve readability overtime since if we only used 1 file it would eventually get very long.~~

~~The first test I decided to add for 3 clients is the simple recursive_find_content finding content 2 leaps away.~~

~~The content is stored in client C and client A is requesting it. Client A only knows client B, and client B only knows client C.~~

~~So hence Client A asks client B then client B asks client C then the data  is returned.~~

kdeme reviewed (approved):
Note: My reviews here apply mostly on the soundness off the added test. I can't give that much feedback on the actual rust code :)

Regarding the test, two remarks:
1. Whether `RecursiveFindContent` actually gets properly tested might depend on some of the initial routing table population strategies. I think it will mostly be fine, because the test is likely to run before that occurs, but it could be the case that node A already knows node C because of a concurrent FindNodes calls launched to populate the routing table at start-up.
2. Depends on the implementation, but a node should normally not return a node that is not yet verified. In fluffy, this will not be a problem because the `portal_historyAddEnr` call will add the ENR and also register it immediatly as verified. However, I just checked and this is not specified in the Portal JSON-RPC API. Something that we should add I think.

ogenev reviewed (changes requested):
I think it will be better if we create a new simulator for tests that include 2+ clients.

The `portal-interop`  name suggests testing interoperability between the portal clients, not exactly more specific scenarios such as gossip or findcontent/nodes which I would call more a network functionality or mini test networks.

Another benefit of creating a separate simulator would be a better UX for the frontend to see what categories of tests are falling and for cli users to just run the portal-interop suite without requiring running all other network tests.  Having a simulator with 100+ tests is not very practical IMO and the cost of creating new simulators is not big so I'm more in the camp of more simulators with fewer tests than one huge simulator.

Blocking this until we decide on a path forward.

KolbyML commented:
Ok so I made a new simulator called ``portal-scenarios``.

And I switched the test from a recursive find content to a find_content test. This time only B knows about an other node which is C. We call findcontent from clientA to B, B returns a list of Enrs to A then A calls find_content again with the enr returned and gets the content from node C.

The reason I switched to this test is because it gives a better picture and is easier to debug then recursive (also a better test since we don't have to worry about A already knowing C)

I do run into an issue though ``Fluffy -> Fluffy -> Fluffy`` works like 98% of the time, but when I run ``Trin -> Trin -> Trin`` it works like 2% of the time and returns around 3 different errors on testing.

Anyways before I can form any solid conclusions I will have to debug it more

ogenev commented:
I'm not sure if I'm in love with the  `portal-scenarios` name, probably something more like `portal-mesh` or dunno, of course, this is an opinionated suggestion.

> I do run into an issue though `Fluffy -> Fluffy -> Fluffy` works like 98% of the time.

I think we want to investigate this further and see if that 2% are actually bugs in Fluffy or an issue with the test. We don't want to introduce flaky tests into the suite.

KolbyML commented:
> I'm not sure if I'm in love with the `portal-scenarios` name, probably something more like `portal-mesh` or dunno, of course, this is an opinionated suggestion.

Then how about `potal-3-client-test-suite`?

> > I do run into an issue though `Fluffy -> Fluffy -> Fluffy` works like 98% of the time.
> 
> I think we want to investigate this further and see if that 2% are actually bugs in Fluffy or an issue with the test. We don't want to introduce flaky tests into the suite.

Of course I just threw out numbers, honestly it could have been 100% after I fix the trin issue I will look into that more.

KolbyML commented:
@ogenev Ok I just renamed it to ``portal-mesh`` we can change it later if someone really wants it, doesn't matter to me.

I added 2 tests since find_content return enrs branches depending if the reciept node knows if other nodes closer then itself. To achieve this I had to add private keys section. It comes from the idea nick said of just passing in a private key to the nodes, as it really is the cleanest solution.

If we put None for the array they are all random before, if we use random as an element in the array that one with None will have a randomly generated enr.

Anyways this PR is ready to review

ogenev reviewed (approved):
LGTM :+1:.

KolbyML commented on clients/fluffy/fluffy.sh:
```diff
@@ -5,4 +5,9 @@ set -e
 
-fluffy --rpc --rpc-address="0.0.0.0" --nat:extip:"$IP_ADDR" --network=none --log-level="debug"
+# if HIVE_PRIVATE_KEY isn't set or doesn't exist do y, else do z
+if [ -z ${HIVE_PRIVATE_KEY+x} ]; then
```
resolved

KolbyML commented on hivesim-rs/src/simulation.rs:
```diff
@@ -18,6 +18,12 @@ impl Default for Simulation {
 }
 
+#[derive(serde::Serialize, serde::Deserialize)]
+struct SimulatorConfig {
```
resolved
